concordium-node/src/bin/cli.rs

```diff
@@ -485,36 +488,104 @@ async fn maybe_do_out_of_band_catchup(
-        let from = download_catchup_files(download_url.clone(), data_dir_path).await;
-        match from {
-            Ok(path) => {
-                info!("Starting out of band catch-up");
-                if let Err(e) = consensus.import_blocks(&path) {
-                    if import_stopped.load(atomic::Ordering::Acquire) {
-                        info!("Out of band catchup stopped.");
-                    } else {
-                        info!("Could not complete out of band catch-up due to: {:#}.", e);
-                    }
-                } else {
-                    info!("Completed out of band catch-up from {}.", download_url);
-                }
-                // attempt to properly clean up the downloaded file.
-                if let Err(e) = path.close() {
-                    error!("Could not delete the downloaded file: {}", e);
-                }
-            }
-            Err(e) => {
-                // report error, but continue startup
-                error!("Downloading catch-up files failed: {}.", e)
+        info!("Starting out of band catch-up");
+        if let Err(e) =
+            import_missing_blocks(consensus, regenesis_arc, download_url, data_dir_path).await
+        {
+            if import_stopped.load(atomic::Ordering::Acquire) {
+                info!("Out of band catchup stopped.");
+            } else {
+                info!("Could not complete out of band catch-up due to: {:#}.", e);
             }
+        } else {
+            info!("Completed out of band catch-up from {}.", download_url)
         }
     }
 }
 
-async fn download_catchup_files(
-    download_url: url::Url,
+// An index entry for a chunk of blocks. Its format must correspond to one
+// produced by `database-exporter`.
+#[derive(serde::Deserialize)]
+struct BlockChunkData {
+    // exported chunk of blocks' filename
+    filename:           String,
+    // genesis block index from which relative heights of blocks in the chunk are counted
+    genesis_index:      usize,
+    // relative height of the oldest block stored in the chunk
+    first_block_height: u64,
+    // relative height of the newest block stored in the chunk
+    last_block_height:  u64,
+}
+
+async fn import_missing_blocks(
+    consensus: &ConsensusContainer,
+    regenesis_arc: Arc<Regenesis>,
+    index_url: &url::Url,
     data_dir_path: &Path,
-) -> anyhow::Result<TempPath> {
+) -> anyhow::Result<()> {
+    let best_block_height = consensus.get_best_block_height();
+    let genesis_block_hashes = regenesis_arc.blocks.read().unwrap();
+    let current_genesis_index = genesis_block_hashes.len() - 1;
+    let current_genesis_block_hash = genesis_block_hashes[current_genesis_index].to_string();
+    let json_genesis_block_value: serde_json::Value =
+        serde_json::from_str(&consensus.get_block_info(&current_genesis_block_hash)?)?;
+    let current_genesis_block_height = json_genesis_block_value["blockHeight"].as_u64().unwrap();
```

Comment:
This wouldn't be necessary if you get the relative block height rather than absolute block height.